### PR TITLE
Added support for regions targeting FrameworkContentElements. 

### PR DESCRIPTION
**Pull request by JimWuerch**
Fixes issue #1321  .

Changes proposed in this pull request:
- DelayedRegionCreationBehavior will now successfully attach to regions targeting FrameworkContentElement also.
- Attempting to call Attach with invalid target elements will cause an ArgumentException

Prior to these changes it was random chance as to whether or not your region got created if it wasn't derived from FrameworkElement.  Since FrameworkContentElement is a parallel class to FrameworkElement generated from the same source, it should also be considered a valid target for a region.  No other changes are needed, as both elements have the same Loaded event that is fired at the same time.

For FrameworkElement targets there will be no performance difference, as the codepath remains unchanged for these targets.

Adding an exception is possibly a breaking change, but  without it, there's a hard-to-debug silent failure, as the only reference to the behavior is held by the Loaded event handler.

### CODE REVIEW

**Comment by spsrbud** [comment]
Adding an exception is definitely a breaking change because now you're forcing every object to be either a FrameworkElement or FrameworkContentElement. This goes against the original design which accepted any object that was a DependencyObject. A good example is Avalon Dock's LayoutDocumentPaneGroup which does not derive from FrameworkElement. This change would prevent that object from becoming a region.

A better solution is to keep the instance of DelayedRegionCreationBehavior alive when it's attached. I made a change to DelayedRegionCreationBehavior.cs which I've attached. I added a static list to the class that will add the instance when it's attached and remove the instance when it's unattached. This static list keeps the object alive without having to depend on FrameworkElement's Loaded event. This change will work for all DependencyObject objects that host a region.
[DelayedRegionCreationBehavior.cs.zip](https://github.com/PrismLibrary/Prism/files/1621282/DelayedRegionCreationBehavior.cs.zip)

File: Source/Wpf/Prism.Wpf/Regions/Behaviors/DelayedRegionCreationBehavior.cs [hunk 1]
```diff
@@ -144,7 +144,18 @@ private void WireUpTargetElement()
...
+
+            FrameworkContentElement fcElement = this.TargetElement as FrameworkContentElement;
+            if (fcElement != null)
+            {
+                fcElement.Loaded += this.ElementLoaded;
+                return;
             }
+
+            throw new ArgumentException(string.Format(CultureInfo.InvariantCulture,
+                                                  Resources.RegionTargetInvalidType));
```
Comment (dvorn) [inline Source/Wpf/Prism.Wpf/Regions/Behaviors/DelayedRegionCreationBehavior.cs]: What is the purpose of string.Format here?  There is no arguments.

**Comment by dvorn** [comment]
It seems reasonable to combine the two approaches: first try to hook into Loaded, and if that not possible, add behavior to tracking list.  Hooking into Loaded has a benefit that the region is created earlier, without waiting for UpdatingRegions event.

@spsrbud Is sync object really necessary?  Attach and detach seem to be executed both on the dispatcher thread.

**Comment by spsrbud** [comment]
@dvorn You're right, it's probably not necessary. I just always try to lock access to a shared resource as a more defensive approach for future enhancements. If you want to take it out I don't mind, but, to play devil's advocate, does it hurt anything by leaving it in there?

**Comment by dvorn** [comment]
@spsrbud Then there's a lot of places in Prism.Regions where we add or remove some items to some collections (e.g., RegionManager.Regions or Region.Views).  Do you feel we need to add synchronization in there?

**Comment by spsrbud** [comment]
@dvorn The examples you provided are instance members, so it's probably ok the way it is. I only added a lock to mine to follow best practices of locking a static variable.

**Comment by dvorn** [comment]
updatingRegionsListeners in RegionManager is static :-)

**Comment by spsrbud** [comment]
Just because it might be safe today doesn't mean it will be safe tomorrow. I would use a lock to guard updatingRegionListeners. What happens if multiple threads call RegionManager.UpdateRegions()? I don't know the answer to that, but there's potential for bad things to happen. Always use a lock when accessing a shared resource.

**Comment by dvorn** [comment]
UpdateRegions is triggered when the user is accessing regions collections in RegionManager. All regions stuff is designed to work from the GUI thread.  Although the user may try to call the API from another thread, in most cases this will result in InvalidOperationException when some properties of DependencyObjects are been modified.

If you do not have a particular scenario where the lock is needed, I think we better continue the tradition of not locking the access.

**Comment by JimWuerch** [comment]
I tried to make the absolute minimum changes, but I'd be happier with a more general solution that works with DependencyObject.  I didn't feel confident enough about my knowledge of Prism to make more than the token changes I have.  

One of the hacks I tried when I was building a private copy of Prism to work around things was to just have the UpdatingRegions event hold a strong reference to the behavior instead of a weak one.  It solves the problem, but it's a lot messier than just having a static list of attached behaviors. 

I do really like having both approaches.  I'm new to submitting pull requests, so I'll make the changes suggested, and read up on how to update the pull request.

**Comment by brianlagunas** [comment]
Hold on now. I wouldn't go making all those changes just yet.  This is getting way outside the scope of what I am comfortable with. Especially since there have been very few reports of this.  I would keep what you have except remove the exception.  If more work is needed it can be discussed in s separate issue with a sample the reproduces the issue.

**Comment by JimWuerch** [comment]
If I remove the exception, should there at least be some sort of debug message or something?   I wouldn't have had to spent much time debugging this if it threw an exception, because the errant behavior would've been immediately obvious.    I feel like the biggest bug here is things fail silently.

**Comment by spsrbud** [comment]
@JimWuerch I opened another issue #1328 related to this one that will fix the issue completely. Hopefully they'll accept it and it will never fail silently again.

**Comment by brianlagunas** [comment]
Thanks for the PR. Its a small risk addition, and I hope it is beneficial to many more people who have not voiced their issues yet.  Thanks again.

**Pull request merged**